### ISSUE TITLE
Investigate replaced Whitehall attachment on discarded edition

### ISSUE THREAD

**Issue by floehopper** [alphagov/asset-manager#543]
When trying to check that the Whitehall attachment metadata had been correctly set in Asset Manager, @chrislo and I came across the following example:

A request to the following URL serves the attachment from Whitehall (i.e. the underlying file system) with a `200 OK` response:

https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/251493/Audit_committees_consultation.pdf

A request to the following URL (after signing in) does a `301 Moved Permanently` redirect to a replacement attachment served from Asset Manager:

https://draft-assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/310014/DH_Consultation_on_NHS_Trust.pdf

Initial investigation suggests that the following has happened in Whitehall:

1. Draft edition created
2. Attachment added
3. Edition published
4. New draft edition created
5. Attachment replaced
6. Draft edition discarded

The reason Whitehall is (correctly) serving the original attachment and not redirecting is because the replacement `AttachmentData` is associated with an edition in the `deleted` state and so [the call to `BaseAttachmentsController#attachment_visible?` from `AttachmentsController#show`][1] (correctly) returns `true`.

The reason Asset Manager is (incorrectly) redirecting to the replacement attachment is that the Asset Manager asset has a replacement and we are making the request via the `draft-assets` host, i.e. [this condition in `WhitehallMediaController#download`][2] is `true`.

One thing that confuses me is that I thought that discarding a draft edition resulted in [all its attachments being marked as deleted][3]. However, the `Attachment` associated with the replacement `AttachmentData` does not appear to be marked as deleted. I don't really know how this can have happened. However, it does at least explain why the replacement attachment asset is not marked as deleted in Asset Manager.

Even if the replacement attachment asset was marked as deleted I think Asset Manager might still attempt to do the redirect, but I think it would raise an exception, because [`asset.replacement` would return `nil` in `BaseMediaController#redirect_to_replacement_for`][4].

[1]: https://github.com/alphagov/whitehall/blob/c3286b3c49bec88a4bebbb8d881f7b4737501292/app/controllers/attachments_controller.rb#L13
[2]: https://github.com/alphagov/asset-manager/blob/9fc82473a1151763cea62791803ff1b66ed4cd91/app/controllers/whitehall_media_controller.rb#L23
[3]: https://github.com/alphagov/whitehall/blob/c3286b3c49bec88a4bebbb8d881f7b4737501292/app/services/edition_deleter.rb#L20
[4]: https://github.com/alphagov/asset-manager/blob/9fc82473a1151763cea62791803ff1b66ed4cd91/app/controllers/base_media_controller.rb#L31-L33

**Comment by floehopper** [comment]
Note that I had a go at extending the relevant integration test to include this scenario in [this branch][1], although I'm not sure it's quite right.

[1]: https://github.com/alphagov/whitehall/compare/add-integration-test-for-discarded-attachment-replacement-scenario

**Comment by floehopper** [comment]
I think there might be an issue with `AttachmentData#deleted?`, because it looks as if it only takes into account whether the edition has been deleted; I think it should also take into account whether the `Attachment` is marked as deleted, i.e. when a single attachment is deleted and the edition is not deleted.

**Comment by floehopper** [comment]
> I think there might be an issue with `AttachmentData#deleted?`, because it looks as if it only takes into account whether the edition has been deleted; I think it should also take into account whether the `Attachment` is marked as deleted, i.e. when a single attachment is deleted and the edition is not deleted.

I don't think the above statement about `AttachmentData#deleted?` is true after all. I think I must've been confused by the mention of "attachable" in the `include_deleted_attachables` option passed to `AttachmentData#significant_attachment`.

In fact, as the name clearly suggests, the latter returns an instance of an `Attachment` and thus `AttachmentData#deleted?` *does* take account of whether the attachment has been marked as deleted.

I thought this was worth clarifying, but note that this doesn't resolve the original problem as described in the issue description.

[1]: https://github.com/alphagov/whitehall/blob/42b0803183ee918a2ae285729e7983c3da095456/app/models/attachment_data.rb#L107-L109

**Comment by elliotcm** [comment]
I've now gone through this myself.

I've followed the steps above on integration and could not reproduce the issue.  Everything in whitehall's DB wound up as you would expect.

I do agree that the attachments listed above ended up in an odd non-deleted state while the edition itself is marked as deleted which appears impossible due to the code linked above which deletes all attachments when the edition is deleted

However, the code for this asset deletion was added 23 March 2016, while the edition in question was discarded 9 May 2014.  The [commit where the code was added](https://github.com/alphagov/whitehall/commit/1832d02d9fcbb45f5eeb8204548597cccbc7b99f) calls this situation out specifically, though it leaves out the asset manager issue as it was presumably not a thing for whitehall at the time.

I'm guessing that when the assets were imported from whitehall to asset manager this discrepancy was copied over accidentally.
